src/validation.cpp

```diff
@@ -2246,17 +2247,30 @@ bool CChainState::FlushStateToDisk(
     {
         bool fFlushForPrune = false;
         bool fDoFullFlush = false;
+
         CoinsCacheSizeState cache_state = GetCoinsCacheSizeState(&m_mempool);
         LOCK(cs_LastBlockFile);
         if (fPruneMode && (fCheckForPruning || nManualPruneHeight > 0) && !fReindex) {
+            // make sure we don't prune below the blockfilterindexes bestblocks
+            // pruning is height-based
+            int index_lowest_blockheight = m_chain.Height();
+            bool has_blockfilter = false;
+            ForEachBlockFilterIndex([&index_lowest_blockheight, &has_blockfilter](BlockFilterIndex& index) {
+                int height = index.GetSummary().best_block_height;
+                if (height < index_lowest_blockheight) {
+                    index_lowest_blockheight = height;
+                }
+                has_blockfilter = true;
+            });
+
             if (nManualPruneHeight > 0) {
                 LOG_TIME_MILLIS_WITH_CATEGORY("find files to prune (manual)", BCLog::BENCH);
 
-                m_blockman.FindFilesToPruneManual(setFilesToPrune, nManualPruneHeight, m_chain.Height());
+                m_blockman.FindFilesToPruneManual(setFilesToPrune, has_blockfilter ? std::min(nManualPruneHeight, index_lowest_blockheight) : nManualPruneHeight, m_chain.Height());
             } else {
                 LOG_TIME_MILLIS_WITH_CATEGORY("find files to prune", BCLog::BENCH);
 
-                m_blockman.FindFilesToPrune(setFilesToPrune, chainparams.PruneAfterHeight(), m_chain.Height(), IsInitialBlockDownload());
+                m_blockman.FindFilesToPrune(setFilesToPrune, has_blockfilter ? std::min(chainparams.PruneAfterHeight(), (uint64_t)index_lowest_blockheight) : chainparams.PruneAfterHeight(), m_chain.Height(), IsInitialBlockDownload());
```

Comment:
In commit "Avoid pruning below the blockfilterindex sync height" (bda1c9fc90a704d1962b16fe4242b5637c930b29)

I don't think just passing a lower `nPruneAfterHeight` argument value is sufficient to avoid pruning here. I think the `index_lowest_blockheight` variable (`last_prune` variable in my previous suggestion) has to be passed as a new argument to `FindFilesToPrune`. If you look at the `FindFilesToPrune` implementation you can see it is not using `nPruneAfterHeight` value at all to compute `nLastBlockWeCanPrune`, which is the value we need to change there.

It's also easy to see this code is wrong because it will have no effect at all if index_lowest_blockheight is bigger than chainparams.PruneAfterHeight, which is 100,000.